### ISSUE TITLE
Major bug in raw-md5-opencl

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#1798]
This was with 7869a019e or so, running Titan X.
```
$ ../run/john 25m_md5.in -form:raw-md5-opencl -mask:?a -min-len=3 -max-len=6 -v:1 --enc=raw
Warning: excessive partial hash collisions detected
Loaded 24917564 password hashes with no different salts (Raw-MD5-opencl [MD5 OpenCL])
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
Error building tables: Loaded hash Idx:9046928, No. of Collosions:2
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
128 bit add overflow. in file:bt_hash_type_128.c, line:32.
Press 'q' or Ctrl-C to abort, almost any other key for status
884209g 0:00:09:29 0.01% (6) (ETA: 2015-12-01 00:30) 1553g/s 1305Mp/s 1305Mc/s 31802TC/s GPU:82°C util:99% fan:50% ##vU~~..##vU~~
Session completed
```
Wall clock time was 9:39 for a total wall clock speed of 1283 Mp/s.

oclHashcat does the exact same task a little faster (7:20, wall clock 8:03) and ends at 1762 Mp/s (wall clock 1538 Mp/s). What's more important, the number of cracked hashes differ: It cracks 883802 of them.

While investigating that, the first thing I noticed was we actually only got 883802 entries in JtR's pot file. Second, I also found that a few hundred hashes **differed** in the pot files, for example:
```
-fea79ec6f3830e5a06542fc802fce78a:wakeup
+fea79ec6f6830e5a06542fc802fce78a:wakeup
          ^
-fa1711bb176806d1f1808dbb0e31018a:linger
+fa1711bb276806d1f1808dbb0e31018a:linger
         ^
```
"-" is JtR **and is incorrect**. "+" is oclHashcat. Position of error varies but it is *always* in second word.

I haven't checked other formats for similar bug.

**Comment by magnumripper** [comment]
Oh BTW I haven't yet tried any other device or driver version. It *could* be a driver bug of some sort.

**Comment by Sayantan2048** [comment]
Please give me the recipe to generate 25m hashes that you're using.

**Comment by magnumripper** [comment]
Not sure where I got them (might be from https://www.korelogic.com/InfoSecSouthwest2012_Ripe_Hashes.html) but you could generate them from rockyou.lst or whatever. That's 14M hashes, if you need more just generate md4 too and use as "uncrackable" ones for md5.

**Comment by magnumripper** [comment]
It seems I can't reproduce this on super.

**Comment by magnumripper** [comment]
I need to revisit that machine I used and see what driver version it had.

**Comment by magnumripper** [comment]
I don't even get those "128 bit add overflow. in file:bt_hash_type_128.c, line:32." at all now - was something changed? I did not use the same hash file now though, maybe the one I used had some odd property.

**Comment by magnumripper** [comment]
With another file, I could reproduce on super:
```
$ rm *.pot
$ ../run/john -form:raw-md5-opencl 25m_md5.in -mask:?a -min-len=3 -max-len=6 -v:1 -pot:nvidia.pot -dev=6   
Using default input encoding: UTF-8
Loaded 25000000 password hashes with no different salts (Raw-MD5-opencl [MD5 OpenCL])
|
Progress is too slow!! trying next table size.

Progress is too slow!! trying next table size.

Progress is too slow!! trying next table size.

Progress is too slow!! trying next table size.

Progress is too slow!! trying next table size.

Progress is too slow!! trying next table size.

Progress is too slow!! trying next table size.

Progress is too slow!! trying next table size.
Press 'q' or Ctrl-C to abort, almost any other key for status
1066846g 0:00:11:19 0.01% (6) (ETA: 2015-12-14 16:09) 1570g/s 1094Mp/s 1094Mc/s 26676TC/s GPU:76°C util:48% fan:36% ##8~~~..##vU~~
Session completed

$ ../run/john -form:raw-md5 25m_md5.in -v:1 -loop:nvidia.pot -pot:cpu.pot -ru:none
Using default input encoding: UTF-8
Loaded 25000000 password hashes with no different salts (Raw-MD5 [MD5 128/128 AVX 4x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
1066728g 0:00:00:02 DONE (2015-10-04 18:54) 371682g/s 371682p/s 371682c/s 13049GC/s pyhtt|..~~~~~~
Session completed

$ sort nvidia.pot >1
$ sort cpu.pot >2
$ diff 1 2
2238c2238
< $dynamic_0$006ce4922072b1e3f8b733347fe1a40b:mandy
---
> $dynamic_0$006ce4922072b1e7f8b733347fe1a40b:mandy
31836c31836
< $dynamic_0$063e076fef1f7a27a56a00c249e0da99:oasker
---
> $dynamic_0$063e076fef1f7f28a56a00c249e0da99:oasker
(...)
```
You can copy the file from ~magnum/PUBLIC/25m_md5.in, it's from that "InfoSecSouthwest" stuff. It contains LOTS of crap (garbage, hexified text etc) which explains why your hash table complains. Still, I get an other output than I got on that other host.

**Comment by Sayantan2048** [comment]
While, 'Progress is too slow!! trying next table size.' is acceptable but 'Error building tables: Loaded hash Idx:9046928, No. of Collosions:2' is not. I'll investigate and fix.

**Comment by magnumripper** [comment]
I now confirmed same problem exist using the Tahiti, so it's not a driver bug. Also, it seems it's the very same candidates that gets the wrong hash recorded. Yet the output from Tahiti and the output from Titan X is not exactly the same.

**Comment by magnumripper** [comment]
I tried testing MD4 as well but for some reason the speed is only 10% of raw-md5?

**Comment by magnumripper** [comment]
This is interesting: The nvidia just stores some hashes with some bits flipped, like this:
```diff
@@ -48355,7 +48355,7 @@
 $dynamic_0$098f1fb4853ebd038e54d615b2b294c5:sumo
 $dynamic_0$098f270a457f304ff35e5363cf248562:ftn
 $dynamic_0$098f3eac13b61c2f39e4e6424d7d4d50:pialme
-$dynamic_0$098f6bcd4621d373cade4e832627b4f6:test  <-- the correct one
+$dynamic_0$098f6bcd4121d373cade4e832627b4f6:test
 $dynamic_0$098f9103aab71a5b9d1eeed57fc8a7e0:rose1
 $dynamic_0$098f930ea46b23e67153dec65639d566:m21487
 $dynamic_0$098fa2eb29a5597822aee23eb9cd529f:/.,1
```

...while the AMD stores the correct one and a number of incorrect ones:
```diff
@@ -48355,7 +48355,24 @@
 $dynamic_0$098f1fb4853ebd038e54d615b2b294c5:sumo
 $dynamic_0$098f270a457f304ff35e5363cf248562:ftn
 $dynamic_0$098f3eac13b61c2f39e4e6424d7d4d50:pialme
+$dynamic_0$098f6bcd4121d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4261d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4521d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46213373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46214373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4621d371cade4e832627b4f6:test
 $dynamic_0$098f6bcd4621d373cade4e832627b4f6:test  <-- the correct one
+$dynamic_0$098f6bcd4621d375cade4e832627b4f6:test
+$dynamic_0$098f6bcd4621d854cade4e832627b4f6:test
+$dynamic_0$098f6bcd4621f473cade4e832627b4f6:test
+$dynamic_0$098f6bcd4622d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46261d73cade4e832627b4f6:test
+$dynamic_0$098f6bcd4641d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4641d374cade4e832627b4f6:test
+$dynamic_0$098f6bcd4651d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4671d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46d21d33cade4e832627b4f6:test
+$dynamic_0$098f6bcd7621d373cade4e832627b4f6:test
 $dynamic_0$098f9103aab71a5b9d1eeed57fc8a7e0:rose1
 $dynamic_0$098f930ea46b23e67153dec65639d566:m21487
 $dynamic_0$098fa2eb29a5597822aee23eb9cd529f:/.,1
```
Even stranger, the above only happens for the "test" hash. All other hashes are just incorrectly reported for AMD just like for nvidia. And this happens only for about 30 hashes out of about a million cracked.

**Comment by Sayantan2048** [comment]
I have found the issue. But I intend to merge hash checking for md4/md5/nt before fixing it. It's just that I don't check full hash on gpu. Only the first two words are checked. I'm more worried about the errors like 'Error building table' which you reported in your first test.

**Comment by Sayantan2048** [comment]
> @@ -48355,7 +48355,24 @@
 $dynamic_0$098f1fb4853ebd038e54d615b2b294c5:sumo
 $dynamic_0$098f270a457f304ff35e5363cf248562:ftn
 $dynamic_0$098f3eac13b61c2f39e4e6424d7d4d50:pialme
+$dynamic_0$098f6bcd4121d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4261d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4521d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46213373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46214373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4621d371cade4e832627b4f6:test
 $dynamic_0$098f6bcd4621d373cade4e832627b4f6:test  <-- the correct one
+$dynamic_0$098f6bcd4621d375cade4e832627b4f6:test
+$dynamic_0$098f6bcd4621d854cade4e832627b4f6:test
+$dynamic_0$098f6bcd4621f473cade4e832627b4f6:test
+$dynamic_0$098f6bcd4622d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46261d73cade4e832627b4f6:test
+$dynamic_0$098f6bcd4641d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4641d374cade4e832627b4f6:test
+$dynamic_0$098f6bcd4651d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd4671d373cade4e832627b4f6:test
+$dynamic_0$098f6bcd46d21d33cade4e832627b4f6:test
+$dynamic_0$098f6bcd7621d373cade4e832627b4f6:test
 $dynamic_0$098f9103aab71a5b9d1eeed57fc8a7e0:rose1
 $dynamic_0$098f930ea46b23e67153dec65639d566:m21487
 $dynamic_0$098fa2eb29a5597822aee23eb9cd529f:/.,1

Is this the result of a single session? What I don't understand in the first place is how 'test' password is generated so many times ?

**Comment by magnumripper** [comment]
Yeah it is from a single session. That hash is also among the test vectors... could that somehow interfere?

**Comment by magnumripper** [comment]
I had a look at my original "25m_md5.in". It turns out that file contained a whole lot of pure garbage, like:
```
00000000000000000000000000000000
00000000000000000000000000000001
00000000000000000000000000000002
00000000000000000000000000000003
00000000000000000000000000000004
00000000000000000000000000000005
00000000000000000000000000000006
00000000000000000000000000000007
00000000000000000000000000000008
00000000000000000000000000000009
0000000000000000000000000000000a
00000000000000000000000000000010
00000000000000000000000000000011
00000000000000000000000000000012
00000000000000000000000000000040
000000000000000000000000000000a4
000000000000000000000000000000c4
000000000000000000000000000000f5
00000000000000000000000000000100
00000000000000000000000000000228
0000000000000000000000000000025d
(...lots more...)
```
Removing all hashes with six or more 0's or f's in a row (there were 14300 of them which is more than 350x more than statistically plausible if I did the math right), plus running with --reject-printable[1] (which rejected 2300 more lines of garbage data) fixed the problem (no more "add overflow"). Just doing either of these things was not enough, I had to do both.

So to sum it up, the "add overflow" part of this issue is probably not something you need to address - it was simply bad input and I reckon a few hash buckets were over-booked or something like that. Even core complained with "Warning: excessive partial hash collisions detected".

[1] The --reject-binary option will reject any hash where the full decoded binary (in this case 16 bytes) passes isprint(), which is highly unlikely for a real hash...

**Comment by magnumripper** [comment]
> I have found the issue. But I intend to merge hash checking for md4/md5/nt before fixing it.

@Sayantan2048 if you won't fix this shortly, can you please point me to how to fix it? I don't like having this bug in here for long.
